### PR TITLE
Add script to convert XXXerr() to XXXerror

### PR DESCRIPTION
**Pull request by richsalz**
As discussed in the comment thread at https://github.com/openssl/openssl/pull/9391#discussion_r305412893, this adds new XXXerror() defines that gets rid of the now-unused function name and deprecates the XXXerr() defines.

This is FYI to give a "taste" of what the conversion would be like. In numeric order, the ones remaining are shown below.

Does the OMC want to do this?  If so, I'll continue with the conversion.
```
14 PROPerr(
16 OBJerr(
18 ASYNCerr(
34 PROVerr(
36 UIerr(
41 OCSPerr(
47 FUNCerr(
49 CONFerr(
50 CRYPTOerr(
50 DSAerr(
51 RANDerr(
57 CTerr(
58 CRMFerr(
60 DHerr(
63 OSSL_STOREerr(
63 PKCS12err(
64 KDFerr(
73 BNerr(
73 SM2err(
77 TSerr(
86 ENGINEerr(
86 PEMerr(
99 DSOerr(
108 BIOerr(
118 PKCS7err(
151 X509err(
184 CMSerr(
185 RSAerr(
195 X509V3err(
283 EVPerr(
295 ASN1err(
323 SSLerr(
470 ECerr(
```

### CODE REVIEW

**Comment by t8m** [comment]
Should there be underscore between MODULE and error to better align with other identifiers? Like BUF_error.

**Comment by mattcaswell** [comment]
I think this is much better. I do have a worry about the pain it will cause us when backporting fixes from master to 1.1.1. One option would be to do this - but delay it until shortly before we release 3.0. Or alternatively maybe it won't be so bad and we should just get on with it. I'm not sure.

**Comment by mattcaswell** [comment]
Travis failure is not relevant.

**Comment by richsalz** [comment]
I'm inclined to agree with @t8m that ```FOO_error``` is better than ```FOOerr```; thoughts?

**Comment by richsalz** [comment]
I'll wait a few days and see what the other committers and OMC think before I do more work on this.
(And if it moves forward, I expect to do the complete solution, including an editing script)

**Comment by richsalz** [comment]
Added commit 08a88216176cc7fec454ebd3f1c4e87b15e51628 which includes the editing script and commit 14fb890f86926733a6b6d9f99d961650157aeb0e which shows what things look like using ```CRPTO_error``` style.

**Comment by beldmit** [comment]
I agree with Tomas about undescore and Matt about delay.

**Comment by richsalz** [comment]
Since #9452 is likely to go in, I have removed variations and made this based purely on that.

File: include/openssl/err.h [hunk 1]
```diff
@@ -107,49 +107,48 @@ typedef struct err_state_st {
...
+#  define RANDerr(f,r) ERR_PUT_error(ERR_LIB_RAND,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define RSAerr(f,r) ERR_PUT_error(ERR_LIB_RSA,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define SM2err(f,r) ERR_PUT_error(ERR_LIB_SM2,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define SSLerr(f,r) ERR_PUT_error(ERR_LIB_SSL,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define SYSerr(f,r) ERR_PUT_error(ERR_LIB_SYS,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define TSerr(f,r) ERR_PUT_error(ERR_LIB_TS,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define UIerr(f,r) ERR_PUT_error(ERR_LIB_UI,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define X509V3err(f,r) ERR_PUT_error(ERR_LIB_X509V3,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define X509err(f,r) ERR_PUT_error(ERR_LIB_X509,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+# endif
```
Comment (levitte) [inline include/openssl/err.h]: Please note that with #9452, you will have to adjust this change

Comment (richsalz) [reply include/openssl/err.h]: Can you give me a sample line?  rebased and pushed BTW.

Comment (levitte) [reply include/openssl/err.h]: The sample lines are what you took away, i.e. the ERR_raise lines.

Comment (richsalz) [reply include/openssl/err.h]: Thanks!  Fixup commit pushed.

File: include/openssl/err.h [hunk 2]
```diff
@@ -108,47 +108,47 @@ typedef struct err_state_st {
...
-#  define SSLerr(f,r) ERR_PUT_error(ERR_LIB_SSL,0,(r),OPENSSL_FILE,OPENSSL_LINE)
-#  define SYSerr(f,r) ERR_PUT_error(ERR_LIB_SYS,0,(r),OPENSSL_FILE,OPENSSL_LINE)
-#  define TSerr(f,r) ERR_PUT_error(ERR_LIB_TS,0,(r),OPENSSL_FILE,OPENSSL_LINE)
-#  define UIerr(f,r) ERR_PUT_error(ERR_LIB_UI,0,(r),OPENSSL_FILE,OPENSSL_LINE)
-#  define X509V3err(f,r) ERR_PUT_error(ERR_LIB_X509V3,0,(r),OPENSSL_FILE,OPENSSL_LINE)
-#  define X509err(f,r) ERR_PUT_error(ERR_LIB_X509,0,(r),OPENSSL_FILE,OPENSSL_LINE)
+#  define ASN1err(f,r) ERR_raise(ERR_LIB_ASN1,(r))
+#  define ASYNCerr(f,r) ERR_raise(ERR_LIB_ASYNC,(r))
+#  define BIOerr(f,r) ERR_raise(ERR_LIB_BIO,(r))
+#  define BNerr(f,r)  ERR_raise(ERR_LIB_RSA,(r))
```
Comment (levitte) [inline include/openssl/err.h]: Typo

File: include/openssl/err.h [hunk 3]
```diff
@@ -108,49 +108,48 @@ typedef struct err_state_st {
...
-# define CMPerr(f,r) ERR_raise(ERR_LIB_CMP,(r))
-# define TSerr(f,r) ERR_raise(ERR_LIB_TS,(r))
-# define HMACerr(f,r) ERR_raise(ERR_LIB_HMAC,(r))
-# define CTerr(f,r) ERR_raise(ERR_LIB_CT,(r))
-# define ASYNCerr(f,r) ERR_raise(ERR_LIB_ASYNC,(r))
-# define KDFerr(f,r) ERR_raise(ERR_LIB_KDF,(r))
-# define SM2err(f,r) ERR_raise(ERR_LIB_SM2,(r))
-# define ESSerr(f,r) ERR_raise(ERR_LIB_ESS,(r))
-# define PROPerr(f,r) ERR_raise(ERR_LIB_PROP,(r))
-# define PROVerr(f,r) ERR_raise(ERR_LIB_PROV,(r))
```
Comment (levitte) [inline include/openssl/err.h]: You know, rearranging the order makes it a bit difficult to check if you got it right...  fortunately, Travis is here to chastise you, so see this as a gentle reminder to add `FUNCerr` back

Comment (richsalz) [reply include/openssl/err.h]: No, FUNCerr was added to master so it doesn't have to come back.
This PR is not intended to compile until all the existing old macros are converted.

File: include/openssl/err.h (hunk 2 above)
Comment (richsalz) [reply include/openssl/err.h]: good eyes.  fixed and force-pushed.

File: include/openssl/err.h (hunk 3 above)
Comment (levitte) [reply include/openssl/err.h]: OK then

**Comment by richsalz** [comment]
Taking this out of WIP. It is reasonable to do the conversion (very) late in the development cycle, so I made it easier to do so. This PR now adds the conversion script, and it shows how the old XXXerr defines will get deprecated, by a hack in err.h that should be removed then the script is run.

**Comment by richsalz** [comment]
I changed the title, removed the WIP.

**Comment by richsalz** [comment]
I'd appreciate feedback on the comments in the script, starting at https://github.com/openssl/openssl/pull/9441/files#diff-101f5c3e883943d0b61879f71fa9d71fR14

**Comment by richsalz** [comment]
I added a script to merge ```XXXerr()``` calls that are broken across two lines.  It think it's feasible to merge this PR now and then, when the project decides to do the conversion, just run the scripts and review their output.

**Comment by richsalz** [comment]
I pushed another commit to reformat some code.  When I do that, I can run the automated scripts and build without errors:
```
; config no-deprecated
; perl -pi -e 's/1 \|\|//;'  include/openssl/err.h 
; git grep -l '[A-Z0-9]err([^)]*$' | xargs perl -pi util/merge-err-lines
; git ls-files | grep '\.c$' | xargs perl -pi util/err-to-raise
; make -sj10
; make test
```

I think this PR, which *does not do the big conversion* should be reviewed and merged.  And then, when the project team is ready, they can run the scripts to do the conversion and merge it.

File: include/openssl/err.h [hunk 4]
```diff
@@ -107,49 +107,49 @@ typedef struct err_state_st {
...
+#  define PROVerr(f, r) ERR_raise_data(ERR_LIB_PROV, (r), NULL)
+#  define RANDerr(f, r) ERR_raise_data(ERR_LIB_RAND, (r), NULL)
+#  define RSAerr(f, r) ERR_raise_data(ERR_LIB_RSA, (r), NULL)
+#  define SM2err(f, r) ERR_raise_data(ERR_LIB_SM2, (r), NULL)
+#  define SSLerr(f, r) ERR_raise_data(ERR_LIB_SSL, (r), NULL)
+#  define SYSerr(f, r) ERR_raise_data(ERR_LIB_SYS, (r), NULL)
+#  define TSerr(f, r) ERR_raise_data(ERR_LIB_TS, (r), NULL)
+#  define UIerr(f, r) ERR_raise_data(ERR_LIB_UI, (r), NULL)
+#  define X509V3err(f, r) ERR_raise_data(ERR_LIB_X509V3, (r), NULL)
+#  define X509err(f, r) ERR_raise_data(ERR_LIB_X509, (r), NULL)
```
Comment (levitte) [inline include/openssl/err.h]: It seems a little unnecessary to make them into `ERR_raise_data`...

Comment (richsalz) [reply include/openssl/err.h]: Well, it's hidden behind a macro.  If I leave it as ```ERR_raise``` that will then get turned into ```ERR_raise_data``` so I figured it was more useful to developers to not have to chase down multiple define's.  But if you want it changed, okay.

Comment (richsalz) [reply include/openssl/err.h]: (By which I mean, be expilcit and say "please change it" :)

Comment (levitte) [reply include/openssl/err.h]: (I only made a comment, not a change request. I'll try to remember to be explicit about that)

Comment (richsalz) [reply include/openssl/err.h]: It's not a big deal; I think we sometimes forget that *discussion* is one of the most important things about a PR. :)

**Comment by richsalz** [comment]
Related to https://github.com/openssl/openssl/pull/9333#issuecomment-528781239 this PR can be merged new, safely, and then the steps in https://github.com/openssl/openssl/pull/9441#issuecomment-522171044 can be done just before one of the pre-official releases.

**Comment by richsalz** [comment]
This is the infrastructure, and includes steps to do the work, for converting the old XXXerr calls.  It's been a month.  Ping for review.

File: crypto/evp/evp_lib.c [hunk 5]
```diff
@@ -60,10 +60,10 @@ int EVP_CIPHER_param_to_asn1(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
...
     } else
         ret = -1;
-    if (ret <= 0)
-        EVPerr(EVP_F_EVP_CIPHER_PARAM_TO_ASN1, ret == -2 ?
-               ASN1_R_UNSUPPORTED_CIPHER :
-               EVP_R_CIPHER_PARAMETER_ERROR);
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_CIPHER_PARAM_TO_ASN1, ASN1_R_UNSUPPORTED_CIPHER);
+    else if (ret <= 0)
+        EVPerr(EVP_F_EVP_CIPHER_PARAM_TO_ASN1, EVP_R_CIPHER_PARAMETER_ERROR);
```
Comment (beldmit) [inline crypto/evp/evp_lib.c]: Sorry for being out of context, but is it OK that here we leave EVPerr(), and in https://github.com/openssl/openssl/pull/9441/files#diff-9f773e58c4cc40339f694534e55f6cb3 we replace EVPerr with ERR_raise()?

File: crypto/evp/evp_lib.c [hunk 6]
```diff
@@ -106,10 +106,10 @@ int EVP_CIPHER_asn1_to_param(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
...
     } else
         ret = -1;
-    if (ret <= 0)
-        EVPerr(EVP_F_EVP_CIPHER_ASN1_TO_PARAM, ret == -2 ?
-               EVP_R_UNSUPPORTED_CIPHER :
-               EVP_R_CIPHER_PARAMETER_ERROR);
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_CIPHER_ASN1_TO_PARAM, EVP_R_UNSUPPORTED_CIPHER);
+    else if (ret <= 0)
+        EVPerr(EVP_F_EVP_CIPHER_ASN1_TO_PARAM, EVP_R_CIPHER_PARAMETER_ERROR);
```
Comment (beldmit) [inline crypto/evp/evp_lib.c]: Similar question.

File: crypto/evp/evp_lib.c (hunk 5 above)
Comment (levitte) [reply crypto/evp/evp_lib.c]: I think these are changes are made to make life easier with `util/merge-err-lines`, as that only handles this kind of split:

``` C
    BLARGerr(BLARG_F_FOO,
             BLARG_R_BAR);
```

File: crypto/evp/evp_lib.c (hunk 6 above)
Comment (levitte) [reply crypto/evp/evp_lib.c]: Same answer

File: crypto/evp/evp_lib.c (hunk 5 above)
Comment (beldmit) [reply crypto/evp/evp_lib.c]: Got it.

**Comment by levitte** [comment]
Merged.

0f17ac26fc552b7f233e8765e6b6a278c9731e2a Add script convert XXerr to ERR_raise
f6aca23e268799380e4e4193789ed96db1ed57f3 Deprecate XXXerr() macros
8c0e76813af6a886f66d3e95b908c6c924d2a6ad Add merge-err-lines script
51ba9ebd09bfa3569fee935ef7063394820cd333 Avoid ?: construct in XXXerr calls

**Comment by richsalz** [comment]
THANK YOU!

**Pull request closed without merging**